android/quest/src/main/java/org/smartregister/fhircore/quest/ui/sdc/qrCode/EditTextQrCodeViewHolderFactory.kt

```diff
@@ -16,110 +16,157 @@
+        newQrResponseItem.answer = listOf(answer)
+        return questionnaireViewItem.copy(questionnaireResponseItem = newQrResponseItem)
       }
     }
 
   fun matcher(questionnaireItem: Questionnaire.QuestionnaireItemComponent): Boolean {
-    val codeableConcept =
-      questionnaireItem.getExtensionByUrl(EXTENSION_URL)?.value as? CodeableConcept
-    return codeableConcept?.coding?.firstOrNull { it.system == EXTENSION_CONTROL_SYSTEM }?.code ==
-      EXTENSION_VALUE
+    return questionnaireItem.getExtensionByUrl(QR_CODE_WIDGET_URL) != null
   }
+}
+
+internal class QrCodeViewItemAdapter(private val qrCodeAnswerChangeListener: QrCodeChangeListener) :
+  ListAdapter<QuestionnaireViewItem, QuestionnaireItemViewHolder>(
+    object : DiffUtil.ItemCallback<QuestionnaireViewItem>() {
+      override fun areItemsTheSame(
+        oldItem: QuestionnaireViewItem,
+        newItem: QuestionnaireViewItem,
+      ): Boolean = areContentsTheSame(oldItem, newItem)
+
+      override fun areContentsTheSame(
+        oldItem: QuestionnaireViewItem,
+        newItem: QuestionnaireViewItem,
+      ): Boolean {
+        val newItemAnswers = newItem.answers.map { (it.value as? StringType)?.value }
+        val oldItemAnswers = oldItem.answers.map { (it.value as? StringType)?.value }
+        return oldItem == newItem &&
+          oldItemAnswers.size == newItemAnswers.size &&
+          newItemAnswers.all { oldItemAnswers.contains(it) }
+      }
+    },
+  ) {
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QuestionnaireItemViewHolder {
+    return EditTextQrCodeItemViewHolderFactory(qrCodeAnswerChangeListener).create(parent)
+  }
+
+  override fun onBindViewHolder(holder: QuestionnaireItemViewHolder, position: Int) {
+    holder.bind(getItem(position))
+  }
+}
+
+internal fun interface QrCodeChangeListener {
+  suspend fun onQrCodeChanged(
+    previous: QuestionnaireResponseItemAnswerComponent?,
+    newAnswer: QuestionnaireResponseItemAnswerComponent?,
+  )
+}
+
+internal val QuestionnaireViewItem.isSetOnceReadOnly: Boolean
+  get() {
+    val qrCodeExtension = questionnaireItem.getExtensionByUrl(QR_CODE_WIDGET_URL)
+    val qrCodeEntryModeValue =
+      qrCodeExtension?.getExtensionByUrl(QR_CODE_ENTRY_MODE_URL)?.value as? StringType
+    return qrCodeEntryModeValue?.value == QrCodeEntryMode.SET_ONCE_READ_ONLY.code
+  }
+
+private const val QR_CODE_WIDGET_URL =
+  "https://github.com/opensrp/android-fhir/StructureDefinition/qr-code-widget"
+private const val QR_CODE_ENTRY_MODE_URL = "qr-code-entry-mode"
 
-  private const val EXTENSION_URL =
-    "https://github.com/opensrp/android-fhir/StructureDefinition/questionnaire-itemControl"
-  private const val EXTENSION_CONTROL_SYSTEM =
-    "https://github.com/opensrp/android-fhir/questionnaire-item-control"
-  private const val EXTENSION_VALUE = "qr_code-widget"
+enum class QrCodeEntryMode(val code: String) {
+  SET_ONCE_READ_ONLY(code = "set-only-readonly"),
+  NORMAL(code = "normal"),
```

Comment:
It'd be easier if the mode is based on boolean states/values with the default/initial state being `false`.